File: javascript/056.js
/**
 * Definition for an interval.
 * function Interval(start, end) {
 *     this.start = start;
 *     this.end = end;
 * }
 */
/**
 * @param {Interval[]} intervals
 * @return {Interval[]}
 */
var merge = function(intervals) {
  intervals.sort(function(a, b) { return a.start - b.start; });
  
  var mergedIntervals = [];
  intervals.forEach(function(interval) {
    if (mergedIntervals.length === 0) {
      mergedIntervals.push(interval);
    } else {
      var lastInterval = mergedIntervals[mergedIntervals.length - 1];
      if (interval.start > lastInterval.end) {
        mergedIntervals.push(interval);
      } else if (interval.end > lastInterval.end) {
        lastInterval.end = interval.end;
      }
    }
  });

  return mergedIntervals;
};

File: ruby/056.rb
# Definition for an interval.
# class Interval
#     attr_accessor :start, :end
#     def initialize(s=0, e=0)
#         @start = s
#         @end = e
#     end
# end

# @param {Interval[]} intervals
# @return {Interval[]}
def merge(intervals)
  intervals.sort! {|a, b| a.start <=> b.start}

  result = []
  intervals.each do |interval|
    if result.length == 0
      result.push(interval)
    else
      last_end = result[-1].end
      if interval.start > last_end
        result.push(interval)
      else
        result[-1].end = interval.end if interval.end >= last_end
      end
    end
  end
  result
end